containers/src/Data/Set/Internal.hs

```diff
@@ -594,6 +595,56 @@ delete = go
 {-# INLINE delete #-}
 #endif
 
+alterF :: (Ord a, Functor f) => (Bool -> f Bool) -> a -> Set a -> f (Set a)
+alterF f k s = fmap choose (f member_)
+  where
+    (member_, inserted, deleted) = case alteredSet k s of
+        Deleted d           -> (True , s, d)
+        Inserted i          -> (False, i, s)
+        DeletedInserted d i -> (True , i, d)
+
+    choose True  = inserted
+    choose False = deleted
+#if __GLASGOW_HASKELL__
+{-# INLINABLE alterF #-}
+#else
+{-# INLINE alterF #-}
+#endif
+
+data AlteredSet a
+    = Deleted (Set a)
+      -- ^ The needle is present in the original set and pointer-equal.
+      -- We return the set where the needle is deleted.
+    | Inserted (Set a)
+      -- ^ The needle is not present in the original set.
+      -- We return the set with the needle inserted.
+    | DeletedInserted (Set a) (Set a)
+      -- ^ The needle is present in the original set (@==@) but not pointer-equal.
+      -- We return both the set where the needle is deleted /and/ the set where
+      -- the needle is inserted.
```

Comment:
These fields should all be strict. The `DeletedInserted` constructor needs to document which is which. `alterF` should document what it does in that case.  How does the behavior in that case compare to that of `Data.Map.alterF`?

Reply:
These fields are intentionally lazy in order to avoid unnecessary allocations when we only test membership – I hope that makes sense.

I haven't really unstood `Data.Map.alterF` so far, but I'll try to take another look.

Reply:
> How does the behavior in that case compare to that of `Data.Map.alterF`?

If I it understand correctly, `Data.Map.alterF` uses `replaceAlong` which doesn't replace the key.

https://github.com/haskell/containers/blob/5ff5247f6ea45f8aee8cba3ac77477df24b9dcf2/containers/src/Data/Map/Internal.hs#L1357-L1365

This is a bit surprising to me, since it works differently than `insert` then.

Reply:
Right; it's not the same as `insert`. Which behavior do we want here? Whatever we do, we should document. I'm not overly attached to strictly following `insert`.

Reply:
> Which behavior do we want here?

Personally, I think it's nice to be consistent with `insert` – just to avoid surprising users. OTOH, skipping the pointer-equality check and re-insertion would simplify the implementation and maybe speed it up a bit too.

Should we ask the `lens`-people what they want – since they will be important users of `alterF` via `At`?

Reply:
For the moment, I think maybe we should carefully document this as being unspecified and go with "don't replace" for now. But asking the lens people for input couldn't hurt.

Reply:
Or just document it as not replacing and accepting the inconsistency....

Reply:
I have opened https://github.com/ekmett/lens/issues/901 to discuss the question.

Reply:
Based on the little feedback I got, I have changed the code to preserve equal elements.